File: frontend/mobile/lib/core/services/hadith_service.dart
import 'package:dio/dio.dart';
import '../network/dio_client.dart';
import '../network/api_endpoints.dart';
import '../../features/hadith/data/models/hadith_model.dart';

class HadithService {
  final DioClient _dioClient;

  HadithService(this._dioClient);

  /// Get all Hadith books
  Future<List<HadithBookModel>> getHadithBooks() async {
    try {
      final response = await _dioClient.get('${ApiEndpoints.hadithBase}/books');
      final data = response.data['data'] as List<dynamic>;
      return data.map((json) => HadithBookModel.fromJson(json)).toList();
    } catch (e) {
      throw Exception('Failed to load hadith books: $e');
    }
  }

  /// Get Hadiths by book name
  Future<List<HadithModel>> getHadithsByBook(
    String bookName, {
    int? limit,
    int? offset,
  }) async {
    try {
      final queryParams = <String, dynamic>{
        'book': bookName,
        if (limit != null) 'limit': limit,
        if (offset != null) 'offset': offset,
      };

      final response = await _dioClient.get(
        '${ApiEndpoints.hadithBase}/hadiths',
        queryParameters: queryParams,
      );

      final data = response.data['data'] as List<dynamic>;
      return data.map((json) => HadithModel.fromJson(json)).toList();
    } catch (e) {
      throw Exception('Failed to load hadiths by book: $e');
    }
  }

  /// Get a specific Hadith by ID
  Future<HadithWithDetailsModel> getHadithById(
    String hadithId, {
    bool includeSanad = false,
    bool includeExplanations = false,
  }) async {
    try {
      final queryParams = <String, dynamic>{
        'include_sanad': includeSanad,
        'include_explanations': includeExplanations,
      };

      final response = await _dioClient.get(
        '${ApiEndpoints.hadithBase}/hadiths/$hadithId',
        queryParameters: queryParams,
      );

      return HadithWithDetailsModel.fromJson(response.data['data']);
    } catch (e) {
      throw Exception('Failed to load hadith: $e');
    }
  }

  /// Get a Hadith by number and book
  Future<HadithWithDetailsModel> getHadithByNumber(
    String hadithNumber,
    String bookName, {
    bool includeSanad = false,
    bool includeExplanations = false,
  }) async {
    try {
      final queryParams = <String, dynamic>{
        'include_sanad': includeSanad,
        'include_explanations': includeExplanations,
      };

      final response = await _dioClient.get(
        '${ApiEndpoints.hadithBase}/hadiths/number/$hadithNumber/book/$bookName',
        queryParameters: queryParams,
      );

      return HadithWithDetailsModel.fromJson(response.data['data']);
    } catch (e) {
      throw Exception('Failed to load hadith by number: $e');
    }
  }

  /// Search Hadiths
  Future<HadithSearchResponse> searchHadiths({
    required String query,
    List<String>? books,
    List<HadithGrade>? grades,
    List<String>? themes,
    String searchType = 'text',
    int limit = 20,
    int offset = 0,
  }) async {
    try {
      final queryParams = <String, dynamic>{
        'q': query,
        'type': searchType,
        'limit': limit,
        'offset': offset,
        if (books != null && books.isNotEmpty) 'books': books.join(','),
        if (grades != null && grades.isNotEmpty)
          'grades': grades.map((g) => g.value).join(','),
        if (themes != null && themes.isNotEmpty) 'themes': themes.join(','),
      };

      final response = await _dioClient.get(
        '${ApiEndpoints.hadithBase}/search',
        queryParameters: queryParams,
      );

      return HadithSearchResponse.fromJson(response.data['data']);
    } catch (e) {
      throw Exception('Failed to search hadiths: $e');
    }
  }

  /// Get search suggestions
  Future<List<String>> getSearchSuggestions(String query) async {
    try {
      final response = await _dioClient.get(
        '${ApiEndpoints.hadithBase}/search/suggestions',
        queryParameters: {'q': query},
      );

      final data = response.data['data'] as List<dynamic>;
      return data.map((e) => e as String).toList();
    } catch (e) {
      throw Exception('Failed to get search suggestions: $e');
    }
  }

  /// Get Hadiths by topic/theme
  Future<HadithTopicResponse> getHadithsByTopic(
    String topic, {
    bool includeRelated = false,
    List<HadithGrade>? grades,
    int limit = 20,
    int offset = 0,
  }) async {
    try {
      final queryParams = <String, dynamic>{
        'include_related': includeRelated,
        'limit': limit,
        'offset': offset,
        if (grades != null && grades.isNotEmpty)
          'grades': grades.map((g) => g.value).join(','),
      };

      final response = await _dioClient.get(
        '${ApiEndpoints.hadithBase}/topics/$topic',
        queryParameters: queryParams,
      );

      return HadithTopicResponse.fromJson(response.data['data']);
    } catch (e) {
      throw Exception('Failed to load hadiths by topic: $e');
    }
  }

  /// Get chapters for a book
  Future<List<HadithChapterModel>> getBookChapters(String bookId) async {
    try {
      final response = await _dioClient.get(
        '${ApiEndpoints.hadithBase}/books/$bookId/chapters',
      );

      final data = response.data['data'] as List<dynamic>;
      return data.map((json) => HadithChapterModel.fromJson(json)).toList();
    } catch (e) {
      throw Exception('Failed to load book chapters: $e');
    }
  }

  /// Get Hadiths by narrator
  Future<List<HadithSearchResultModel>> getHadithsByNarrator(
    String narratorName, {
    int limit = 20,
    int offset = 0,
  }) async {
    try {
      return await searchHadiths(
        query: narratorName,
        searchType: 'narrator',
        limit: limit,
        offset: offset,
      ).then((response) => response.results);
    } catch (e) {
      throw Exception('Failed to load hadiths by narrator: $e');
    }
  }
}

/// Response for Hadith search queries
class HadithSearchResponse {
  final List<HadithSearchResultModel> results;
  final int totalCount;
  final String query;
  final String searchType;
  final int searchTimeMs;

  const HadithSearchResponse({
    required this.results,
    required this.totalCount,
    required this.query,
    required this.searchType,
    required this.searchTimeMs,
  });

  factory HadithSearchResponse.fromJson(Map<String, dynamic> json) {
    return HadithSearchResponse(
      results: (json['results'] as List<dynamic>)
          .map((e) => HadithSearchResultModel.fromJson(e))
          .toList(),
      totalCount: json['total_count'] as int,
      query: json['query'] as String,
      searchType: json['search_type'] as String,
      searchTimeMs: json['search_time_ms'] as int,
    );
  }
}

/// Response for topic-based queries
class HadithTopicResponse {
  final String topic;
  final List<HadithWithDetailsModel> hadiths;
  final List<String> relatedTopics;
  final int totalCount;

  const HadithTopicResponse({
    required this.topic,
    required this.hadiths,
    required this.relatedTopics,
    required this.totalCount,
  });

  factory HadithTopicResponse.fromJson(Map<String, dynamic> json) {
    return HadithTopicResponse(
      topic: json['topic'] as String,
      hadiths: (json['hadiths'] as List<dynamic>)
          .map((e) => HadithWithDetailsModel.fromJson(e))
          .toList(),
      relatedTopics: (json['related_topics'] as List<dynamic>)
          .map((e) => e as String)
          .toList(),
      totalCount: json['total_count'] as int,
    );
  }
}

File: frontend/nextjs-app/src/lib/services/hadith-service.ts
import { axiosClient } from '../api/axios-client';
import type {
  Hadith,
  HadithBook,
  HadithChapter,
  HadithWithDetails,
  HadithSearchResponse,
  HadithTopicResponse,
  HadithSearchFilters,
  HadithGrade,
} from '@/types/hadith';

const HADITH_BASE_URL = '/api/v1';

export class HadithService {
  /**
   * Get all Hadith books
   */
  static async getHadithBooks(): Promise<HadithBook[]> {
    try {
      const response = await axiosClient.get(`${HADITH_BASE_URL}/books`);
      return response.data.data;
    } catch (error) {
      console.error('Failed to fetch hadith books:', error);
      throw new Error('Failed to load hadith books');
    }
  }

  /**
   * Get Hadiths by book name
   */
  static async getHadithsByBook(
    bookName: string,
    limit?: number,
    offset?: number
  ): Promise<Hadith[]> {
    try {
      const params: Record<string, any> = { book: bookName };
      if (limit) params.limit = limit;
      if (offset) params.offset = offset;

      const response = await axiosClient.get(`${HADITH_BASE_URL}/hadiths`, { params });
      return response.data.data;
    } catch (error) {
      console.error('Failed to fetch hadiths by book:', error);
      throw new Error('Failed to load hadiths');
    }
  }

  /**
   * Get a specific Hadith by ID
   */
  static async getHadithById(
    hadithId: string,
    includeSanad: boolean = false,
    includeExplanations: boolean = false
  ): Promise<HadithWithDetails> {
    try {
      const params = {
        include_sanad: includeSanad,
        include_explanations: includeExplanations,
      };

      const response = await axiosClient.get(
        `${HADITH_BASE_URL}/hadiths/${hadithId}`,
        { params }
      );
      return response.data.data;
    } catch (error) {
      console.error('Failed to fetch hadith:', error);
      throw new Error('Failed to load hadith details');
    }
  }

  /**
   * Get a Hadith by number and book
   */
  static async getHadithByNumber(
    hadithNumber: string,
    bookName: string,
    includeSanad: boolean = false,
    includeExplanations: boolean = false
  ): Promise<HadithWithDetails> {
    try {
      const params = {
        include_sanad: includeSanad,
        include_explanations: includeExplanations,
      };

      const response = await axiosClient.get(
        `${HADITH_BASE_URL}/hadiths/number/${hadithNumber}/book/${bookName}`,
        { params }
      );
      return response.data.data;
    } catch (error) {
      console.error('Failed to fetch hadith by number:', error);
      throw new Error('Failed to load hadith');
    }
  }

  /**
   * Search Hadiths
   */
  static async searchHadiths(
    query: string,
    filters?: HadithSearchFilters,
    limit: number = 20,
    offset: number = 0
  ): Promise<HadithSearchResponse> {
    try {
      const params: Record<string, any> = {
        q: query,
        type: filters?.searchType || 'text',
        limit,
        offset,
      };

      if (filters?.books && filters.books.length > 0) {
        params.books = filters.books.join(',');
      }

      if (filters?.grades && filters.grades.length > 0) {
        params.grades = filters.grades.join(',');
      }

      if (filters?.themes && filters.themes.length > 0) {
        params.themes = filters.themes.join(',');
      }

      const response = await axiosClient.get(`${HADITH_BASE_URL}/search`, { params });
      return response.data.data;
    } catch (error) {
      console.error('Failed to search hadiths:', error);
      throw new Error('Failed to search hadiths');
    }
  }

  /**
   * Get search suggestions
   */
  static async getSearchSuggestions(query: string): Promise<string[]> {
    try {
      const response = await axiosClient.get(`${HADITH_BASE_URL}/search/suggestions`, {
        params: { q: query },
      });
      return response.data.data;
    } catch (error) {
      console.error('Failed to get search suggestions:', error);
      return [];
    }
  }

  /**
   * Get Hadiths by topic/theme
   */
  static async getHadithsByTopic(
    topic: string,
    includeRelated: boolean = false,
    grades?: HadithGrade[],
    limit: number = 20,
    offset: number = 0
  ): Promise<HadithTopicResponse> {
    try {
      const params: Record<string, any> = {
        include_related: includeRelated,
        limit,
        offset,
      };

      if (grades && grades.length > 0) {
        params.grades = grades.join(',');
      }

      const response = await axiosClient.get(
        `${HADITH_BASE_URL}/topics/${topic}`,
        { params }
      );
      return response.data.data;
    } catch (error) {
      console.error('Failed to fetch hadiths by topic:', error);
      throw new Error('Failed to load hadiths by topic');
    }
  }

  /**
   * Get chapters for a book
   */
  static async getBookChapters(bookId: string): Promise<HadithChapter[]> {
    try {
      const response = await axiosClient.get(`${HADITH_BASE_URL}/books/${bookId}/chapters`);
      return response.data.data;
    } catch (error) {
      console.error('Failed to fetch book chapters:', error);
      throw new Error('Failed to load book chapters');
    }
  }

  /**
   * Get Hadiths by narrator
   */
  static async getHadithsByNarrator(
    narratorName: string,
    limit: number = 20,
    offset: number = 0
  ): Promise<HadithSearchResponse> {
    try {
      return await this.searchHadiths(
        narratorName,
        { searchType: 'narrator' },
        limit,
        offset
      );
    } catch (error) {
      console.error('Failed to fetch hadiths by narrator:', error);
      throw new Error('Failed to load hadiths by narrator');
    }
  }
}
